client/lib/html-toolbar/index.js

```diff
@@ -0,0 +1,229 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import { get, noop, reduce } from 'lodash';
+
+/**
+ * Return an object containing the textarea content split into:
+ * - `before` selection
+ * - (`inner`) selection
+ * - `after` selection
+ *
+ * @param {Element} textarea The editor textarea.
+ * @returns {Object} The split textarea content.
+ */
+export const splitSelectedContent = textarea => {
+	const { selectionEnd, selectionStart, value } = textarea;
+	return {
+		before: value.substring( 0, selectionStart ),
+		inner: value.substring( selectionStart, selectionEnd ),
+		after: value.substring( selectionEnd, value.length ),
+	};
+};
+
+/**
+ * Set the cursor position after the insertion of new content into a textarea.
+ *
+ * @param {Element} textarea The editor textarea.
+ * @param {Number} selectionEnd The textarea.selectionEnd value BEFORE the content insertion.
+ * @param {Number} insertedContentLength The length of the inserted content.
+ */
+export const setCursorPosition = ( textarea, selectionEnd, insertedContentLength ) => {
+	textarea.selectionEnd = textarea.selectionStart = selectionEnd + insertedContentLength;
+};
+
+/**
+ * Insert content into a textarea.
+ *
+ * @param {Element} textarea The target textarea.
+ * @param {String} before The content before the selection.
+ * @param {String} content The new content to insert.
+ * @param {String} after The content after the selection.
+ * @param {Function} onInsert A function to call after the insertion.
+ */
+export const insertContent = ( textarea, before, content, after, onInsert = noop ) => {
+	const { selectionEnd, value } = textarea;
+	const newContent = before + content + after;
+	const userAgent = get( window, 'navigator.userAgent', '' );
+
+	if ( /(?:firefox|fxios)/i.test( userAgent ) ) {
+		// In Firefox, execCommand( 'insertText' ), needed to preserve the undo stack,
+		// always moves the cursor to the end of the content.
```

Comment:
are we sure this is true when we have a selection already?

when I worked with it I though for sure that I could insert text without having the cursor jump to the very end. could it be only when no explicit selection is selected?

do we have a link to a bug report or a behavioral explanation we could add as a comment here?